datalad/plugin/addurls.py

```diff
@@ -891,26 +884,60 @@ def __call__(dataset, urlfile, urlformat, filenameformat,
             log_progress(lgr.info, "addurls_versionurls", "Finished versioning URLs")
 
         files_to_add = set()
-        for r in add_urls(rows, ifexists=ifexists, options=annex_options):
-            if r["status"] == "ok":
-                files_to_add.add(r["path"])
-            yield r
+        # We need to group by dataset since otherwise we will initiate
+        # batched annex process per each subdataset, which might be infeasible
+        # in any use-case with a considerable number of subdatasets.
+        # TODO: logical spot for parallel run across (sub)datasets
+        for subpath, ds_rows in itertools.groupby(rows, itemgetter("subpath")):
```

Comment:
eh Yarik -- RTFM https://docs.python.org/3/library/itertools.html#itertools.groupby "Generally, the iterable needs to already be sorted on the same key function.".  That is why current overall slowness and progressbars virtually never being present -- just keeps adding only few files at a time... not good

Reply:
I changed the rows to be sorted in ba05a61535.  Please of course change/drop that if that's not the right fix.

Reply:
Thanks! unfortunately `itemgetter` is not be enough (not that I figured it out, the script has crashed ;)) since for top level dataset it would be `None` and sorted would fail, thus we either need a tuned up getter or store empty string not None

Reply:
Sorry, I should have considered that (or at least noted the failing tests).

> thus we either need a tuned up getter or store empty string not None

I've updated the key function in 9d8e58d.  Tests pass for me locally.

Reply:
THANK YOU! I will give it a shot